core/server/common/src/main/java/alluxio/master/StateLockManager.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master;
+
+import alluxio.collections.ConcurrentHashSet;
+import alluxio.conf.PropertyKey;
+import alluxio.conf.ServerConfiguration;
+import alluxio.exception.ExceptionMessage;
+import alluxio.resource.LockResource;
+import alluxio.util.ThreadFactoryUtils;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Provides graceful and interruptable locking protocol for taking the state lock.
+ *
+ * {@link #lockShared()} will be used by user RPCs and may throw {@link InterruptedException}
+ * based on options passed to {@link #lockExclusive(StateLockOptions)}.
+ *
+ * {@link #lockExclusive(StateLockOptions)} will be used by metadata backups in order to
+ * guarantee paused state during critical tasks.
+ */
+public class StateLockManager {
+  private static final Logger LOG = LoggerFactory.getLogger(StateLockManager.class);
+
+  /** The state-lock */
+  private ReadWriteLock mStateLock = new ReentrantReadWriteLock(true);
+
+  /** The set of threads that are waiting for state-lock in shared mode. */
+  private Set<Thread> mSharedWaiters;
+  /** The set of threads that are holding the state-lock in shared mode. */
+  private Set<Thread> mSharedHolders;
+
+  /** Scheduler that is used for interrupt-cycle. */
+  private ScheduledExecutorService mScheduler;
+  /** The future for the active interrupt cycle */
+  private ScheduledFuture<?> mInterrupterFuture;
+
+  /** Whether state-lock manager is in safe-mode. */
+  private boolean mSafeMode;
+
+  /**
+   * Creates a new state-lock manager.
+   */
+  public StateLockManager() {
+    mSharedWaiters = new ConcurrentHashSet<>();
+    mSharedHolders = new ConcurrentHashSet<>();
+
+    mScheduler = Executors
+        .newSingleThreadScheduledExecutor(ThreadFactoryUtils.build("state-lock-manager-%d", true));
+
+    mSafeMode = ServerConfiguration.getBoolean(PropertyKey.MASTER_BACKUP_STATE_LOCK_SAFE_MODE);
```

Comment:
does safemode never get reset? That means the master can ONLY take a backup and never service clients, even when it is primary?

Reply:
We can make it something that resets but need to coordinate this between masters.
Right now it's just a crude failsafe.